File: src/SISe.c
/*
 *  siminf, a framework for stochastic disease spread simulations
 *  Copyright (C) 2015  Pavol Bauer
 *  Copyright (C) 2015  Stefan Engblom
 *  Copyright (C) 2015  Stefan Widgren
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* Compartments */
enum {S, I};

/* Offsets in data to parameters in the model */
enum {PHI,
      UPSILON,
      GAMMA,
      ALPHA,
      BETA_Q1,
      BETA_Q2,
      BETA_Q3,
      BETA_Q4,
      EPSILON};

/**
 * susceptible to infected: S -> I
 *
 * @param u The compartment state vector in node.
 * @param t Current time.
 * @param data The data vector for node.
 * @param sd The sub-domain of node.
 * @return propensity.
 */
double SISe_S_to_I(const int *u, double t, const double *data, int sd)
{
    return data[UPSILON] * data[PHI] * u[S];
}

/**
 *  infected to susceptible: I -> S
 *
 * @param u The compartment state vector in node.
 * @param t Current time.
 * @param data The data vector for node.
 * @param sd The sub-domain of node.
 * @return propensity.
 */
double SISe_I_to_S(const int *u, double t, const double *data, int sd)
{
    return data[GAMMA] * u[I];
}

/**
 * Update infectious pressure
 *
 * @param u The compartment state vector in node.
 * @param node The node.
 * @param t Current time.
 * @param data The data vector for node.
 * @param sd The sub-domain of node.
 * @return 1 if needs update, else 0.
 */
int SISe_post_time_step(
    const int *u,
    int node,
    double t,
    double *data,
    int sd)
{
    const int days_in_year = 365;
    const int days_in_quarter = 91;

    double S_n, I_n;
    double tmp = data[PHI];

    S_n = u[S];
    I_n = u[I];

    /* Time dependent beta for each quarter of the year. Forward Euler step. */
    switch (((int)t % days_in_year) / days_in_quarter) {
    case 0:
        data[PHI] *= (1.0 - data[BETA_Q1]);
        break;
    case 1:
        data[PHI] *= (1.0 - data[BETA_Q2]);
        break;
    case 2:
        data[PHI] *= (1.0 - data[BETA_Q3]);
        break;
    default:
        data[PHI] *= (1.0 - data[BETA_Q4]);
        break;
    }

    if ((I_n + S_n) > 0.0)
        data[PHI] += data[ALPHA] * I_n / (I_n + S_n) + data[EPSILON];
    else
        data[PHI] += data[EPSILON];

    /* 1 if needs update */
    return tmp != data[PHI];
}

File: R/SISe.R
## This file is part of SimInf, a framework for stochastic
## disease spread simulations.
##
## Copyright (C) 2015 Pavol Bauer
## Copyright (C) 2017 -- 2019 Robin Eriksson
## Copyright (C) 2015 -- 2019 Stefan Engblom
## Copyright (C) 2015 -- 2026 Stefan Widgren
##
## SimInf is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## SimInf is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see <https://www.gnu.org/licenses/>.

##' Class SISe
##'
##' Class to handle the \acronym{SISe} model. This class inherits from
##' \code{\linkS4class{SimInf_model}}, meaning that \acronym{SISe}
##' objects are fully compatible with all generic functions defined
##' for \code{SimInf_model}, such as \code{\link{run}},
##' \code{\link{plot}}, \code{\link{trajectory}}, and
##' \code{\link{prevalence}}.
##'
##' @template SISe-details
##'
##' @seealso
##' \code{\link{SISe}} for creating an \acronym{SISe} model object and
##' \code{\linkS4class{SimInf_model}} for the parent class definition.
##' @include SimInf_model.R
##' @export
setClass("SISe", contains = c("SimInf_model"))

##' The compartments in an SISe model
##' @noRd
compartments_SISe <- function() {
    compartments_SIS()
}

##' Select matrix for events in the SISe model
##'
##' Internal function returning the 2x2 select matrix (E) that maps
##' SISe compartments (rows) to event types (columns) for event
##' processing.
##'
##' @return A 2x2 numeric matrix with compartments as rows and event
##'     types as columns. Used internally by SimInf_events.
##' @noRd
select_matrix_SISe <- function() {
    select_matrix_SIS()
}

##' Create an SISe model
##'
##' Create an \acronym{SISe} model to be used by the simulation
##' framework.
##'
##' @template SISe-details
##' @details
##' The argument \code{u0} must be a \code{data.frame} with one row
##' for each node with the following columns:
##' \describe{
##' \item{S}{The number of susceptible in each node}
##' \item{I}{The number of infected in each node}
##' }
##'
##' @template u0-param
##' @template tspan-param
##' @template events-param
##' @template phi-param
##' @param upsilon Indirect transmission rate of the environmental
##'     infectious pressure
##' @param gamma A numeric vector with the recovery rate from infected
##'     to susceptible.  Each node can have a different gamma
##'     value. The vector must have length 1 or \code{nrow(u0)}. If
##'     the vector has length 1 but the model contains more nodes, the
##'     value is repeated for all nodes.
##' @template alpha-param
##' @template beta-end-param
##' @param epsilon The background environmental infectious pressure
##' @return \code{SISe}
##' @seealso
##' \code{\linkS4class{SISe}} for the class definition.
##' \code{\link{SIR}}, \code{\link{SEIR}}, \code{\link{SIS}},
##' \code{\link{SISe3}} and \code{\link{SISe_sp}} for other predefined
##' models.  \code{\link{mparse}} for creating custom models.
##' \code{\link{run}} for running the simulation.
##' \code{\link{trajectory}}, \code{\link{prevalence}} and
##' \code{\link{plot,SimInf_model-method}} for post-processing and
##' visualization.
##' @include check_arguments.R
##' @export
SISe <- function(u0,
                 tspan,
                 events  = NULL,
                 phi     = NULL,
                 upsilon = NULL,
                 gamma   = NULL,
                 alpha   = NULL,
                 beta_t1 = NULL,
                 beta_t2 = NULL,
                 beta_t3 = NULL,
                 beta_t4 = NULL,
                 end_t1  = NULL,
                 end_t2  = NULL,
                 end_t3  = NULL,
                 end_t4  = NULL,
                 epsilon = NULL) {
    ## Check arguments.

    ## Check u0 and compartments
    u0 <- check_u0(u0, compartments_SISe())

    ## Check initial infectious pressure
    if (is.null(phi))
        phi <- 0
    phi <- rep(phi, length.out = nrow(u0))
    check_infectious_pressure_arg(nrow(u0), phi)

    ## Check for non-numeric parameters
    check_gdata_arg(upsilon, gamma, alpha, beta_t1, beta_t2, beta_t3, beta_t4,
                    epsilon)

    ## Check interval endpoints
    check_integer_arg(end_t1, end_t2, end_t3, end_t4)
    end_t1 <- rep(end_t1, length.out = nrow(u0))
    end_t2 <- rep(end_t2, length.out = nrow(u0))
    end_t3 <- rep(end_t3, length.out = nrow(u0))
    end_t4 <- rep(end_t4, length.out = nrow(u0))
    check_end_t_arg(nrow(u0), end_t1, end_t2, end_t3, end_t4)

    ## Arguments seem ok...go on

    G <- matrix(c(1, 1, 1, 1), nrow = 2, ncol = 2,
                dimnames = list(c("S -> upsilon*phi*S -> I",
                                  "I -> gamma*I -> S"),
                                c("1", "2")))

    S <- matrix(c(-1,  1, 1, -1), nrow = 2, ncol = 2,
                dimnames = list(compartments_SISe(), c("1", "2")))

    v0 <- matrix(as.numeric(phi), nrow  = 1, byrow = TRUE,
                 dimnames = list("phi"))

    ldata <- matrix(as.numeric(c(end_t1, end_t2, end_t3, end_t4)),
                    nrow  = 4, byrow = TRUE,
                    dimnames = list(c("end_t1", "end_t2", "end_t3", "end_t4")))

    gdata <- as.numeric(c(upsilon, gamma, alpha, beta_t1, beta_t2,
                          beta_t3, beta_t4, epsilon))
    names(gdata) <- c("upsilon", "gamma", "alpha", "beta_t1", "beta_t2",
                      "beta_t3", "beta_t4", "epsilon")

    model <- SimInf_model(G      = G,
                          S      = S,
                          E      = select_matrix_SISe(),
                          tspan  = tspan,
                          events = events,
                          ldata  = ldata,
                          gdata  = gdata,
                          u0     = u0,
                          v0     = v0)

    methods::as(model, "SISe")
}
